payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java

```diff
@@ -359,6 +359,16 @@ private PriorPaymentControlResult getPluginPurchaseResult(final PaymentControlCo
                 return new DefaultPriorPaymentControlResult(true);
             }
 
+            final List<InvoicePayment> existingInvoicePayments = invoiceApi.getInvoicePaymentsByInvoice(invoiceId, internalContext);
+            for (final InvoicePayment existingInvoicePayment : existingInvoicePayments) {
```

Comment:
Would it make sense to optimize and skip any `isSuccess` `existingInvoicePayment ` (to avoid fetching all associated payment transactions in that case?

Reply:
I don't think it's worth the optimization: 99.99% of the time, `existingInvoicePayments` is empty.